CP-44752: propagate System.Diagnostics tracing information using W3C traceparent header

edwintorok opened the pull request:
The newer System.Net.Http would propagate this automatically, but the older WebRequest doesn't.

Retrieve the span identifier and add the header ourselves. The WebRequest object is fresh for each request, so setting the headers shouldn't cause race conditions.

This adds a dependency on a library that is part of .Net source repository, but distributed outside of it on NuGet: System.Diagnostics. Lock the dependency to a given hash to avoid accidental/malicious changes.
This dependency is not available on .Net45, so all the code and build dependencies are wrapped with appropriate conditional compilation directives (which may look kind of ugly, they can be dropped once XenServer.dll drops support for .Net45 and switches to .Net462 minimum).

No new instrumentation is created, so if the application using XenServer.dll is not configured/set up to use OpenTelemetry or System.Diagnostics.DiagnosticSource, then this is a no-op.

I've tested this using the Powershell 7 module.

Draft PR because:
* I haven't tested the version with the locked dependencies yet
* There is one TODO in the code about obtaining a version number

edwintorok commented:
Internal build succeeded with locked package dependencies too.

edwintorok commented:
I think I'll also need `NET462_OR_GREATER` otherwise XC won't pick this up.

danilo-delbusso commented on ocaml/sdk-gen/csharp/autogen/src/JsonRpc.cs:
```diff
@@ -155,6 +158,11 @@ public partial class JsonRpcClient
         private int _globalId;
 
+#if (NET462_OR_GREATER || NETSTANDARD2_0_OR_GREATER)
+        // TODO: get proper version string from build
```
We never merge source with TODOs on `master`.

Use the name of the package `XenServer.NET` and its version. I think we can get it from https://learn.microsoft.com/en-us/dotnet/api/system.reflection.assemblyname.version?view=netstandard-2.0, but needs testing

If not, this needs a placeholder, and changes to pipelines. We can advise on that if needed

kc284 replied:
If the client code uses the C# SDK as a project in their code, `typeof(JsonRpcClient).Assembly.GetName().Version` should return the SDK version, but in the corner case of someone dropping the sources in a client project (like XenCenter does), it should return the client project version (in XC it would return XenModel's version, which is XC's version).

If we can't have the ocaml build use a placeholder and replace it with the tag (we ran into some trouble with @psafont when we tried it for the API reference), I suggest using `new ActivitySource("XenAPI.JsonRpcClient", "@SDK_VERSION@")` (the truth is I dislike placeholders like this, but we have one more insance of it in the PS code).

By the way, is it indeed the SDK version we need here? (As opposed to a string that will be representing the implementing client, somehting like the UserAgent, for example).

danilo-delbusso replied:
> but in the corner case of someone dropping the sources in a client project (like XenCenter does)


we could do without the placeholder and just add a line in the README :shrug:

danilo-delbusso replied:
> By the way, is it indeed the SDK version we need here? (As opposed to a string that will be representing the implementing client, somehting like the UserAgent, for example).

for this, I'd say so

if a client wants to use open telemetry, it should add it itself and then use its own user agent

I may be wrong - I'm not that familiar with what the expected behaviour is

kc284 replied:
> it should add it itself and then use its own user agent

Yes, the thing is we offer the option to use own UserAgent (albeit it sub-optimally designed in my opinion), but we don't offer this option for the activity source with this design.

edwintorok replied:
https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Api/README.md#instrumenting-a-libraryapplication-with-net-activity-api and https://learn.microsoft.com/en-us/dotnet/core/diagnostics/distributed-tracing-instrumentation-walkthroughs say 
> Create an ActivitySource, providing the name and version of the library/application doing the instrumentation
> . Each library or library subcomponent can (and often should) create its own source. Consider creating a new source rather than reusing an existing one if you anticipate app developers would appreciate being able to enable and disable the Activity telemetry in the sources independently.
> The source name passed to the constructor has to be unique to avoid the conflicts with any other sources. If there are multiple sources within the same assembly, use a hierarchical name that contains the assembly name and optionally a component name, for example, Microsoft.AspNetCore.Hosting. If an assembly is adding instrumentation for code in a second, independent assembly, the name should be based on the assembly that defines the ActivitySource, not the assembly whose code is being instrumented.

So this is for developers to identify the origin of this instrumentation, in this case XenServer.dll. So even if the application as a whole wants to use another user-agent, I think using a constant name is fine here, because we need to be able to find the source code of what created this instrumentation.

In fact we should be able to determine this from the runtime, I'll try this:
```
+        private static readonly System.Reflection.AssemblyName AssemblyName = typeof(JsonRpcClient).Assembly?.GetName();
+        private static ActivitySource source = new ActivitySource(AssemblyName?.FullName, AssemblyName?.Version?.ToString());
```

danilo-delbusso commented on ocaml/sdk-gen/csharp/autogen/src/JsonRpc.cs:
```diff
@@ -225,6 +241,10 @@ protected virtual T Rpc<T>(string callName, JToken parameters, JsonSerializer se
                     using (var responseStream = new MemoryStream())
                     {
+#if (NET462_OR_GREATER || NETSTANDARD2_0_OR_GREATER)
+                        // https://opentelemetry.io/docs/specs/semconv/rpc/json-rpc/
```
please add a comment explaining the tag within the source. links can change and not work anymore. If you still think it'd be useful to have a link, I'd use https://archive.org/

danilo-delbusso commented on ocaml/sdk-gen/csharp/autogen/src/JsonRpc.cs:
```diff
@@ -256,14 +289,25 @@ protected virtual T Rpc<T>(string callName, JToken parameters, JsonSerializer se
-                                        if (errorArray != null)
+                                        if (errorArray != null) {
+#if (NET462_OR_GREATER || NETSTANDARD2_0_OR_GREATER)
+                                            activity?.SetTag("otel.status_code", "ERROR");
```
I'd use `const` for all these strings here

edwintorok replied:
We could use the constants from here: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/src/Shared/SemanticConventions.cs
There was a discussion upstream about releasing that as a separate NuGet package, but I don't see that yet.
For now I'll copy the constants that I use from there, so if we want to switch to using that NuGet package in the future the constant names would already match.

danilo-delbusso reviewed (changes requested):
haven't tested it yet but I spotted a couple of things on the first run

for SDK PRs also please request reviews manually, because we don't monitor the repo as often as toolstack does, so it helps to see PR reviews under https://github.com/pulls

kc284 commented on ocaml/sdk-gen/csharp/autogen/src/JsonRpc.cs:
```diff
@@ -256,14 +289,25 @@ protected virtual T Rpc<T>(string callName, JToken parameters, JsonSerializer se
+                                        if (errorArray != null) {
+#if (NET462_OR_GREATER || NETSTANDARD2_0_OR_GREATER)
+                                            activity?.SetTag("otel.status_code", "ERROR");
+                                            //activity?.SetTag("rpc.jsonrpc.error_message", errorArray);
```
Should we be imlementing this properly rather than commenting it out - currently there is disparity between v1 and v2 implementation.

edwintorok replied:
added ErrorMessage. Code may or may not be present, so I didn't set that one. We shouldn't be using JSONRPC v1 anymore anyway.

kc284 commented on ocaml/sdk-gen/csharp/autogen/src/JsonRpc.cs:
```diff
@@ -239,12 +259,25 @@ protected virtual T Rpc<T>(string callName, JToken parameters, JsonSerializer se
 #endif
+
+#if (NET462_OR_GREATER || NETSTANDARD2_0_OR_GREATER)
+                                    activity?.SetTag("rpc.jsonrpc.version", "2.0");
```
Shouldn't we be setting this tag further down for v1 too?

edwintorok replied:
> Conditionally Required If other than the default version (1.0)

I'll add it just to be explicit though.

kc284 commented on ocaml/sdk-gen/csharp/autogen/src/JsonRpc.cs:
```diff
@@ -293,6 +337,23 @@ protected virtual void PerformPostRequest(Stream postStream, Stream responseStre
+            // propagate W3C traceparent and tracestate
+            // HttpClient would do this automatically on .NET 5,
+            // and .NET 6 would provide even more control over this: https://blog.ladeak.net/posts/opentelemetry-net6-httpclient
+            // the caller must ensure that the activity is in W3C format (by inheritance or direc setting)
```
typo: direc->direct

kc284 reviewed (dismissed):
As above.

edwintorok commented:
I've addressed the review comments, and tested a build, the `traceparent` is sent correctly:
```
POST /jsonrpc HTTP/1.1
Host: lcy2-dt72.xenrt.citrite.net
Accept: application/json
User-Agent: XenServerPSModule/25.2.0.0
Connection: Keep-Alive
traceparent: 00-a29a037790a77a956ab8ce8887d1287d-900ac8ca433ee4f4-01
Content-Type: application/json
Content-Length: 100
```

edwintorok commented:
I'll add a small self-contained example here on how the output looks with OpenTelemetry instrumentation enabled.

edwintorok commented:
<img width="1892" alt="Screenshot 2025-01-27 at 11 51 51" src="https://github.com/user-attachments/assets/3f33569a-f233-4f8b-b85c-81197a399865" />
<img width="1718" alt="Screenshot 2025-01-27 at 11 51 41" src="https://github.com/user-attachments/assets/a4672e68-3a17-4dc1-b1b4-44013a2ab946" />

Tested using a slightly tweaked version of the above sample (using OTLP to send to jaeger), and confirmed that we get a joined Client + Server trace in Jaeger.

mg12 reviewed (approved):
The code changes and output looks good to me, and the joined Client + Server trace in Jaeger shows it's working as intended. Perhaps someone more used to the SDK build system could check the dependency changes @kc284 @danilo-delbusso

edwintorok commented:
It has been discussed that we should switch to .Net 8, so I'll see if I can add an additional build target for that here (that'll ensure that the right DLL gets picked out of the Nuget).

edwintorok commented:
The xenserver-samples fail to run:
```
\r\nException: System.IO.FileNotFoundException: Could not load file or assembly &apos;System.Diagnostics.DiagnosticSource, Version=8.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51&apos;. The system cannot find the file specified.\r\nFile name: &apos;System.Diagnostics.DiagnosticSource, Version=8.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51&apos;\r\n   at XenAPI.JsonRpcClient.Rpc[T](String callName, JToken parameters, JsonSerializer serializer)\r\n   at XenAPI.JsonRpcClient.session_login_with_password(String _uname, String _pwd, String _version, String _originator)\r\n   at XenAPI.Session.login_with_password(String username, String password, String version, String originator)\r\n   at Citrix.XenServer.Commands.ConnectXenServerCommand.ProcessRecord()\r\n   at System.Management.Automation.CommandProcessor.ProcessRecord()\r\n</log>\r\n</test>\r\n<test>\r\n<name>Create SR create_nfs_sr 10.63.96.18 /mia11_export/4288734-QzbMXt PowerShellAutoTestSR True</name>\r\n<state>Fail</state>\r\n<log>\r\nCmd: 'create_nfs_sr 10.63.96.18 /mia11_export/4288734-QzbMXt PowerShellAutoTestSR3jigbrbkso6'\r\nException: System.Exception: Could not find open sessions to any XenServers.\r\n   at System.Management.Automation.MshCommandRuntime.ThrowTerminatingError(ErrorRecord errorRecord)\r\n</log>\r\n</test>\r\n<test>\r\n<name>Install VM install_vm PowerShellAutoTestVM PowerShellAutoTestSR True</name>\r\n<state>Fail</state>\r\n<log>\r\nCmd: 'install_vm PowerShellAutoTestVM42q95qmrdwg PowerShellAutoTestSR3jigbrbkso6'\r\nException: System.Exception: Could not find open sessions to any XenServers.\r\n   at System.Management.Automation.MshCommandRuntime.ThrowTerminatingError(ErrorRecord errorRecord)\r\n</log>\r\n</test>\r\n<test>\r\n<name>Start VM start_vm PowerShellAutoTestVM Running</name>\r\n<state>Fail</state>\r\n<log>\r\nCmd: 'start_vm PowerShellAutoTestVM42q95qmrdwg'\r\nException: System.Exception: Could not find open sessions to any XenServers.\r\n   at System.Management.Automation.MshCommandRuntime.ThrowTerminatingError(ErrorRecord errorRecord)\r\n</log>\r\n</test>\r\n<test>\r\n<name>Shutdown VM shutdown_vm PowerShellAutoTestVM Halted</name>\r\n<state>Fail</state>\r\n<log>\r\nCmd: 'shutdown_vm PowerShellAutoTestVM42q95qmrdwg'\r\nException: System.Exception: Could not find open sessions to any XenServers.\r\n   at System.Management.Automation.MshCommandRuntime.ThrowTerminatingError(ErrorRecord errorRecord)\r\n</log>\r\n</test>\r\n<test>\r\n<name>Uninstall VM uninstall_vm &apos;PowerShellAutoTestVM&apos; True</name>\r\n<state>Fail</state>\r\n<log>\r\nCmd: 'uninstall_vm &apos;PowerShellAutoTestVM42q95qmrdwg&apos;'\r\nException: System.Exception: Could not find open sessions to any XenServers.\r\n   at System.Management.Automation.MshCommandRuntime.ThrowTerminatingError(ErrorRecord errorRecord)\r\n</log>\r\n</test>\r\n<test>\r\n<name>Destroy SR detach_nfs_sr PowerShellAutoTestSR True</name>\r\n<state>Fail</state>\r\n<log>\r\nCmd: 'detach_nfs_sr PowerShellAutoTestSR3jigbrbkso6'\r\nException: System.Exception: Could not find open sessions to any XenServers.\r\n   at System.Management.Automation.MshCommandRuntime.ThrowTerminatingError(ErrorRecord errorRecord)\r\n</log>\r\n</test>\r\n<test>\r\n<name>Disconnect Server</name>\r\n<state>Pass</state>\r\n<log />\r\n</test>\r\n</group>\r\n</results>\r\n" (```.

Maybe the dependencies don't get propagated correctly from the SDK to the app, although I would've expected Nuget to do the right thing.

I'll try another build where I *only* enable tracing when you're on .Net 8.0+ and not in other cases (that means that XenCenter likely won't get it, and we'll also lack tracing in Powershell -- which was an easy way to test that the tracing works).

The C# sample fails to run too, but that probably needs the environment updated to have .Net 8.0 available (I switched the sample from 6.0 to 8.0):
```
[09:39:28 23/04] SSH root@10.63.101.170 chmod +x csharp/XenSdkSample && csharp/XenSdkSample 'https://10.63.97.46' root ************ (10.63.97.46 = genus-35-095d, 10.63.101.170 = linux_debian@genus-35-095d (debian12))
You must install .NET to run this application.

App: /root/csharp/XenSdkSample
Architecture: x64
App host version: 8.0.13
.NET location: Not found

Learn more:
https://aka.ms/dotnet/app-launch-failed

Download the .NET runtime:
https://aka.ms/dotnet-core-applaunch?missing_runtime=true&arch=x64&rid=linux-x64&os=debian.12&apphost_version=8.0.13```

edwintorok commented:
With the latest changes here (enable tracing *only* for .NET 8.0), and updating the tests to install the .Net 8 runtime when testing the .Net 8 sample -- the tests have passed, so I think this is ready.

edwintorok commented:
@kc284 could you take another look at this?

kc284 commented on Makefile:
```diff
@@ -83,6 +83,11 @@ doc:
 sdk:
 	dune build --profile=$(PROFILE)  @sdkgen xapi-sdk.install @ocaml/sdk-gen/install
 
+# workaround for no .resx generation, just for compilation testing
```
I'm not sure why you need these lines here for running the githb workflows.
